be/src/formats/parquet/file_writer.cpp

```diff
@@ -177,139 +192,718 @@ void ParquetBuildHelper::build_compression_type(::parquet::WriterProperties::Bui
+        return _add_int_column_chunk<TYPE_UNSIGNED_INT, ::parquet::Type::INT32>(ctx, type_desc, node, col);
+    }
+    case TYPE_BIGINT: {
+        return _add_int_column_chunk<TYPE_BIGINT, ::parquet::Type::INT64>(ctx, type_desc, node, col);
+    }
+    case TYPE_UNSIGNED_BIGINT: {
+        return _add_int_column_chunk<TYPE_UNSIGNED_BIGINT, ::parquet::Type::INT64>(ctx, type_desc, node, col);
+    }
+    case TYPE_CHAR:
+    case TYPE_VARCHAR: {
+        return _add_varchar_column_chunk(ctx, type_desc, node, col);
+    }
+    case TYPE_DATETIME: {
+        return _add_datetime_column_chunk(ctx, type_desc, node, col);
+    }
+    case TYPE_DATE: {
+        return _add_date_column_chunk(ctx, type_desc, node, col);
+    }
+    case TYPE_DECIMAL32: {
+        return _add_decimal_column_chunk<TYPE_DECIMAL32, ::parquet::Type::INT32>(ctx, type_desc, node, col);
+    }
+    case TYPE_DECIMAL64: {
+        return _add_decimal_column_chunk<TYPE_DECIMAL64, ::parquet::Type::INT64>(ctx, type_desc, node, col);
+    }
+    case TYPE_DECIMAL128: {
+        return _add_decimal_column_chunk<TYPE_DECIMAL128, ::parquet::Type::FIXED_LEN_BYTE_ARRAY>(ctx, type_desc, node,
+                                                                                                 col);
+    }
+    case TYPE_STRUCT: {
+        return _add_struct_column_chunk(ctx, type_desc, node, col);
+    }
+    case TYPE_ARRAY: {
+        return _add_array_column_chunk(ctx, type_desc, node, col);
+    }
+    case TYPE_MAP: {
+        return _add_map_column_chunk(ctx, type_desc, node, col);
+    }
+    default: {
+        return Status::InvalidArgument(fmt::format("Type {} is not supported", type_desc.debug_string()));
+    }
+    }
+}
+
+Status FileWriterBase::_add_struct_column_chunk(Context& ctx, const TypeDescriptor& type_desc,
+                                                const ::parquet::schema::NodePtr& node, const ColumnPtr& col) {
+    DCHECK(type_desc.type == TYPE_STRUCT);
+    auto group_node = std::static_pointer_cast<::parquet::schema::GroupNode>(node);
+
+    unsigned char* nulls = nullptr;
+    if (col->is_nullable()) {
+        const auto null_column = down_cast<NullableColumn*>(col.get())->null_column();
+        nulls = null_column->get_data().data();
+    }
+
+    const auto data_column = ColumnHelper::get_data_column(col.get());
+    const auto struct_column = down_cast<StructColumn*>(data_column);
+    Context child_ctx(ctx._max_def_level + node->is_optional(), ctx._max_rep_level, ctx.size() + data_column->size());
+
+    int subcol_size = 0;
+    auto def_delta = node->is_optional();
+    for (auto i = 0; i < ctx.size(); i++) {
+        auto [idx, def_level, rep_level] = ctx.get(i);
+        if (idx == Context::kNULL) {
+            child_ctx.append(Context::kNULL, def_level, rep_level);
+            continue;
+        }
+        if (nulls != nullptr && nulls[idx]) {
+            child_ctx.append(Context::kNULL, def_level, rep_level);
+            subcol_size++;
```

Comment:
why increment `subcol_size` if it's null?

Reply:
Null entry in StructColumn occupies a position in sub-columns.